include/pybind11/pybind11.h

```diff
@@ -2611,13 +2611,23 @@ void print(Args &&...args) {
 }
 
 error_already_set::~error_already_set() {
-    if (m_type) {
-        gil_scoped_acquire gil;
-        error_scope scope;
-        m_type.release().dec_ref();
-        m_value.release().dec_ref();
-        m_trace.release().dec_ref();
-    }
+    // Not using py::gil_scoped_acquire here since that calls get_internals,
+    // which is known to trigger failures in the wild (a full explanation is
+    // currently unknown).
+    // gil_scoped_acquire_local here is a straight copy from get_internals code.
+    // I.e. py::gil_scoped_acquire acquires the GIL in detail/internals.h exactly
+    // like we do here now, releases it, then acquires it again in gil.h.
+    // Using gil_scoped_acquire_local cuts out the get_internals overhead and
+    // fixes the failures observed in the wild. See PR #1895 for more background.
+    struct gil_scoped_acquire_local {
+        gil_scoped_acquire_local() : state(PyGILState_Ensure()) {}
+        ~gil_scoped_acquire_local() { PyGILState_Release(state); }
+        const PyGILState_STATE state;
+    } gil;
+    error_scope scope;
+    m_type.release().dec_ref();
```

Comment:
@rwgk m_type, m_value etc could be null after having been moved from, shouldn't we add the if check back in?

Reply:
Oh, I didn't think of that, thanks!
Yes you're right. I'll also add a test.

Reply:
I added a copy & move tests. It turns out, we're fine. What makes this work is the use of `Py_XDECREF`:
```
    const handle &dec_ref() const & {
        Py_XDECREF(m_ptr);
        return *this;
    }
```

I believe given that, the current code is actually ideal: even if there is some odd mix of `m_type`, `m_value`, `m_trace` still holding a reference, it will correctly release the resources.

I ran the new test with ASAN, MSAN, TSAN, and it is also going through valgrind in the CI that's still running.

I verified with `printf` in the `error_already_set` dtor that `// Both destructors run.` is in fact true.

Reply:
Ugh ... this just came in:
```
/__w/pybind11/pybind11/tests/test_exceptions.cpp:312:48: error: std::move of the const variable 'caught' has no effect; remove std::move() or make the variable non-const [performance-move-const-arg,-warnings-as-errors]
                py::error_already_set moved_to{std::move(caught)};
                                               ^~~~~~~~~~      ~
/__w/pybind11/pybind11/tests/test_exceptions.cpp:315:35: error: local copy 'copied_to' of the variable 'caught' is never modified; consider avoiding the copy [performance-unnecessary-copy-initialization,-warnings-as-errors]
            py::error_already_set copied_to{caught};
                                  ^
            const                &
6193 warnings generated.
```

Of course ... looking again

Reply:
Fixed, and everything seems to be happy.

Reply:
@rwgk Yeah, I know we are making using of Py_XDECREF etc, but and if statement would be more efficient, no? Instead of relying on the implicit if statements in all of the object types + unnecessary copy of nullptrs in release.

Reply:
@rwgk More importantly, we won't incur the overhead from error_scope and locking and releasing the GIL for no reason.